[#40] 주문 프로세스를 위한 파사드 계층에 RestTemplate 세팅

choichanhyeok opened the pull request:
## a. 설명
> 주문 프로세스를 위한 파사드 세팅

## b. 작업 내용
> RestTemplate와 RabbitMQ
- 재고 감소 요청과 결제 요청은 동기적으로 처리해야 하기에 RestTemplate 활용키로
- 배송 요청의 경우 비동기적 처리를 위해 RabbitMQ 활용키로

> @Configuration 클래스에 restTemplate를 빈으로 등록해 주입받을 수 있도록
- restTemplate를 이용해 AOP 같은 스프링 컨테이너의 관리가 필요한 작업에 사용할게 아닌데 이런 경우에 빈에 등록하는게 어떤 이점이 있는지는 모르겠으나 대부분의 예시 코드들이 이렇게 작성되어 있어 차용함.

## c. 작업 회고
> 주문 서비스와 상품 서비스 사이의 재고 감소를 위한 Dto 동기화
- 각 서비스 별로 dto가 필요할텐데 이런 dto의 변경은 일반적으로 중간에서 관리해주는 무언가가 있는건지
- 중간에서 요청 응답을 하는 큐 같은게 따로 필요한건지

> RestTemplate와 WebClient
- WebClient를 써도 옵션에 따라 동기적으로 처리할 수 있음을 확인
- 그렇지만 WebFlux를 사용하지 않는 환경, 블록킹 I/O가 발생해도 문제가 될 만큼의 트래픽이 없을 것으로 가정했을 때 RestTemplate가 적합하다고 판단함.

> Kafka와 RabbitMQ
- 현재는 단순 메시지 전송 역할이기에 경량 메시지 브로커인 RabbitMQ가 적합하다고 판단해 적용키로함

choichanhyeok committed:
[#40] 주문 프로세스를 위한 파사드 계층에 RestTemplate 세팅

- 재고 감소, 결제 요청의 경우 동기 처리가 필요하기에 RestTemplate 이용키로
- 배송 요청의 경우 RabbitMQ를 이용해 비동기 처리 계획
- RestTemplate는 별도 config에 bean을 등록해 주입받을 수 있도록 함

f-lab-troy reviewed (approved):
- Facade에 대해 이해하기
- OrderService에 processOrder등을 넣는 것에 대한 것 고려해보기

choichanhyeok committed:
[#40] 재고 감소, 결제, 배송 요청등의 작업을 파사드가 아닌 OrderService에서 직접 처리토록

- OrderFacade 제거
- 재고 감소, 결제, 배송요청을 위한 RESTTEMPLATE 요청을 OrderService에서 진행토록 변경
- 변경 이후 기존 테스트 코드가 정상 동작토록 재정의
- RestTemplate를 빈에 등록해 사용키 위해 Config에 등록

choichanhyeok commented:
아래 이유로 기존 Facade 계층을 지우고 OrderService의 `createOrder`에서 재고 감소, 결제, 배송 등록 서비스에 대한 `RestTemplate`요청을 진행토록 수정했습니다.

> 이유 1. 파사드가 필요할 정도로 OrderService에 createOrder이 복잡해지는가
```
위 두 로직 모두 결국 restTemplate로 상품, 주문 도메인에 요청을 던지는 부분이기에 그렇게 복잡한 로직이 아니라고 생각한다.
또 현재 createOrder()에서 다루는 코드도 크게 복잡하지 않아 바로 Service로 들어와도 흐름을 파악하는게 어렵지 않을 것 같다.
```

<br>

> 이유 2. 나중에 필요성을 느낄 때 수정하는게 어려운 문제인가
```
재고 감소 요청이나 결제 과정에서 실패가 일어날 경우 이에 대한 처리를 해주는 부분이 복잡해 질 수 있을 거 같으나 
그 때 필요한 클래스를 생성하고 메서드를 분리하면 될 거 같다.
```

<br>

> 이유 3. OrderService의 createOrder()에서 주문 프로세스 전반을 처리하는 경우 SRP와 DIP를 위반하는가
```
RestTemplate 요청도 결국 해당 객체를 통해 하기 때문에 SRP를 충족한다고 생각한다. 
차후 예외처리가 필요할 경우 관련 클래스나 메서드를 분리해 작업토록 하면 이후에도 SRP를 유지할 수 있을 거 같다.

DIP의 경우도 RestTemplate를 빈으로 등록해 주입받기 때문에 준수하고 있다고 생각한다.
```
